File: Benchmarks/4.1-calculos-matematicos/prueba-1/version-1/python/main.py
import gc
import time
import random
import js  # type: ignore
import tracemalloc
from pyscript import display


def create_matrix(size):
    return [[random.random() for _ in range(size)] for _ in range(size)]


def multiply_matrices(size):
    tracemalloc.start()
    A = create_matrix(size)
    B = create_matrix(size)
    C = [[0] * size for _ in range(size)]

    start = time.time()
    gc.collect()
    for i in range(size):
        for j in range(size):
            _sum = 0
            for k in range(size):
                _sum += A[i][k] * B[k][j]
            C[i][j] = _sum

    end = time.time()

    execution_time = (end - start) * 1000
    display(f"ET: {round(execution_time, 2)} ms", target="pyscript-output")

    js.endTimerWebAssembly()

    memory_usage = tracemalloc.get_traced_memory()[1] / (1024 * 1024)
    display(f"RAM: {round(memory_usage, 2)} MB",
            target="pyscript-output")
    tracemalloc.stop()


def run_py_benchmark(event):
    js.clearCell('pyscript-output')
    multiply_matrices(300)

File: Benchmarks/4.1-calculos-matematicos/prueba-1/version-1/javascript/main.js
function runNodeBenchmark() {
    clearCell("node-js-output");
    fetch("http://localhost:3000/")
        .then(response => response.json())
        .then(data => {
            let outputDiv = document.getElementById("node-js-output");

            let timeDiv = document.createElement("div");
            timeDiv.textContent = `ET: ${data.executionTime} ms`;
            outputDiv.appendChild(timeDiv);

            let cpuDiv = document.createElement("div");
            cpuDiv.textContent = `CPU: ${data.cpuUsage} %`;
            outputDiv.appendChild(cpuDiv);

            let memoryDiv = document.createElement("div");
            memoryDiv.textContent = `RAM: ${data.memoryUsage} MB`;
            outputDiv.appendChild(memoryDiv);

        })
        .catch(error => console.error("Error:", error));
}

function createMatrix(size) {
    let matrix = new Array(size);
    for (let i = 0; i < size; i++) {
        matrix[i] = new Array(size);
        for (let j = 0; j < size; j++) {
            matrix[i][j] = Math.random();
        }
    }
    return matrix;
}

function multiplyMatrices(size) {
    const A = createMatrix(size);
    const B = createMatrix(size);
    const C = Array.from({ length: size }, () => new Array(size).fill(0));

    const start = performance.now();

    for (let i = 0; i < size; i++) {
        for (let j = 0; j < size; j++) {
            let sum = 0;
            for (let k = 0; k < size; k++) {
                sum += A[i][k] * B[k][j];
            }
            C[i][j] = sum;
        }
    }

    const executionTime = performance.now() - start;
    const memoryUsage = getMemoryUsageJS();

    const results = {
        execution_time: executionTime,
        memory_usage: memoryUsage
    };

    displayResults(results);
}

function getMemoryUsageJS() {
    if (performance.memory) {
        return performance.memory.usedJSHeapSize / (1024 * 1024);
    }
    return -1;
}

function displayResults(results) {
    const output = document.getElementById("javascript-output");

    const timeDiv = document.createElement("div");
    timeDiv.textContent = `ET: ${results.execution_time.toFixed(2)} ms`;
    output.appendChild(timeDiv);

    const memoryDiv = document.createElement("div");
    if (results.memory_usage !== -1) {
        memoryDiv.textContent = `RAM: ${results.memory_usage.toFixed(2)} MB`;
    } else {
        memoryDiv.textContent = `RAM unsupported measurement in this browser.`;
    }
    output.appendChild(memoryDiv);
}

function runJsBenchmark(event) {
    clearCell("javascript-output");
    multiplyMatrices(300);
}
